rfd/0150-ssh-connection-resumption.md

```diff
@@ -0,0 +1,119 @@
+
+### Handshake (new connection)
+
+The clients sends `\x00` (likely together with the previous data), and the server generates a _resumption token_ out of 16 random bytes, but with the highest bit of the first byte set. The server sends this resumption token, then its length-prefixed host ID (which is almost always a UUID in string form, but can also be an EC2 instance ID of varying length). The host ID of the server should be used in later reconnection attempts to ensure that the correct server is receiving the new connection, since the first connection might've been to a hostname or public address and thus might eventually resolve to a different server.
+
+### Handshake (reconnection)
+
+The client sends the previously-received _resumption token_, to which the server replies with a `\x01` to confirm that the token is for a connection that can be resumed or `\x00` if the token is unknown (or for a connection that's expired) followed by a connection close. In the latter case, the client should stop further reconnection attempts and treat the connection as closed by the remote side. The extra byte to confirm or deny the validity of the token is necessary to signal to the client that the connection should be treated as invalid - closing the connection will just trigger a new reconnection attempt.
+
+After the acknowledgement, both sides will send a count of how many bytes they have received in the connection.
+
+### Data exchange
+
+After both sides agree that the connection can proceed, data is exchanged in frames consisting of an integer count of how many bytes are being acknowledged as successfully received (since the previous acknowledgement, or since the beginning of the connection, to reduce the size of the variable-length integer), then a length-prefixed chunk of data. The size of the chunk must not exceed a size of 128KiB. If either side cannot proceed from the exchanged resume point, or if either side wants to explicitly close the connection, a value of `0xffff_ffff_ffff_ffff` should be sent (as the acknowledgement). Upon receiving such value, the connection should be considered closed by the remote side.
+
+### State
+
+Both sides need to maintain a log of data that has been generated by the application and potentially sent over the wire, but hasn't yet been acknowledged by the peer. Such a buffer should be limited in size to avoid unconstrained resource consumption - from some early tests, a buffer size of 2MiB seems to be sufficient to maintain near full bandwidth when transferring data across a single SSH channel, compared to a regular SSH connection that doesn't make use of the resumption protocol. The protocol does not mandate the use of a buffer with a fixed maximum size, so other approaches can be used in the future if need be.
```

Comment:
Some time ago I was investigating if we can align  quic-go  flowcontraol and / ackhandle https://github.com/quic-go/quic-go/tree/master/internal/ackhandler in TCP protocol instead of UDP but this seesm to be not possible. 

Nevertheless writing our own  ackhandler and flowcontrol will is quite risky and we should investigate already existing solution before starting building our own ACK stack.

Reply:
We're not going to handle flow control ourselves - I tried getting in the same sort of multiplexing that's built in to SSH with a fixed window size, but it required a really big buffer on the receive side to maintain performance, which is pretty bad.

The approach I have tested interleaves acks of received data and outbound data, which does have the potential to deadlock if both sides end up blocked on write without the ability to read, but thankfully SSH _needs_ constant interleaved reads and writes - and, as an extreme failsafe, the deadlock will be resolved after a transport timeout and subsequent reconnection. 😇

Reply:
It is something to think about if we were to extend this to other protocols that are less forgiving than SSH tho.

Reply:
I tried using `https://github.com/hashicorp/yamux` for the wire protocol, with one stream just for data and one stream for acks, and I got similar performance to a previous version of the bespoke wire protocol that used receive windows, but with really big buffers (16MiB) on either end. I am fairly confident that **for our own `x/crypto/ssh`-based SSH server** we won't hit deadlocks if we alternate data and acks without any receive window management (as that's effectively baked into the SSH protocol anyway).